### DISCUSSION TITLE
Reflections about automation and having a good changelog

### DISCUSSION THREAD

**Discussion by vaab** [olivierlacan/keep-a-changelog#466] [Tools]
Here are my 2 cents,

First, I'm the author of `gitchangelog` which will "dump git logs into your changelog". But I feel somewhat that we are chasing the same rabbit.

Can you tell me what's wrong with this changelog: https://pypi.python.org/pypi/gitchangelog#changelog

TBH, I have difficulties to see what can't be automated if you accurately move to the commit time the editing/filtering job of making a changelog. This requires some conventions (and thus discipline) although. But has huge advantages also.

With little effort on your git committing conventions, I think I could generate your actual `CHANGELOG.md` from `git log` thanks to a `gitchangelog` with a customized `gitchangelog.rc`.

I would be happy to contribute to your project in my way, by providing a template in `gitchangelog` that would generate what you came up as what should be the standard format for a changelog.

I feel like you imply that there are no way to generate automatically a valid changelog from a git history, and this seems like a religious statement as I can't see any clear points in your README.md linking the automation process to are failure of following what you list as "what makes a good changelog".

Many thanks to read me, I care about making good changelog also.
I would be happy to have any feedback about my thoughts and hope that mine is helping in some ways.

**Comment by elyscape** [comment]
The problem is that this requires the git commit messages to be designed for the change log and annotated for parsing into one.  Git commit messages [are best used to explain what the change does and (crucially) how the change was made](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).  To quote this project's site:

> The purpose of a commit is to document one atomic step in the process by which the code evolves from one state to another. The purpose of a change log is to document the noteworthy differences between these states.
> 
> As is the difference between good comments and the code itself, so is the difference between a change log and the commit log: one describes the _why_, the other the how.

In other words, git commit messages will in most cases involve detailed plumbing that doesn't belong in a change log.  Rather than require developers to write commit messages with information suitable for a change log, why not instead just require them to document their changes in one directly?

This is even more relevant if your project is on GitHub or Bitbucket or otherwise accepts public activity, as you can be guaranteed that many of your external contributors won't bother to format their commit messages to whatever schema you use.  Often, the commit messages of external contributions are next to worthless.  If you generate your change log from that, you're going to have to do a bunch of editing anyway.  You could just reject all changes with commit messages that don't work for you, but then you're cutting a substantial portion of your potential contributor base.  Manually maintaining a change log provides a middle ground that lets you do the editing piecemeal as contributions drift in, and if you're lucky maybe some contributors will even include changes in the log as part of their pull requests.

**Comment by vaab** [comment]
Thanks, you've answered my question. And I had missed this paragraph about the distinction of "how" and "why". I usually refer to this as the "targeted audience", and we could say that the user is the target audience of the changelog and relates to the "why". And when explaining "how" we are targeting the developers.
However I feel that this "how/why" frame fall short quite quickly. For instance, devs want to know why you made a refactor (and not "how"), and users don't want to know about it (speaking of a refactor having no consequences for user). To be added to that, I feel "how" is often answered by the code itself. However it might need sometimes some more explanation that could be left in the code comments or in the body of your commit message... if you made clear that this message was meant only for the developer to be read.

But it's true that sometimes, you might need to speak to both audiences (or speak about the "why" and the "how" with your perspective). And this case, my current git commit message conventions are falling short to give a proper solution. Although, when this will arise as a real issue, I could probably think about a hopefully not so intrusive way to annotate part of your commit body message by targeted audience.
Another place where my strategy fall short is the maximum commit summary message length. In various place, this message is used by various tools and a general de-facto optimum 50-max character convention is stemming out of current usage. And 50-char is way to small for a proper head changelog entry. 

Again, this could also be tackled by a non-intrusive convention if needed.

Speaking of enforcing conventions, however, I don't feel like your point make sense as it seems very similar to enforce a commit convention than a changelog convention. And yes, I often have to rewrite commit message from external contributors, or have to remind them about conventions, but my projects at each commit are clean and complete, they can generate their changelog and I don't need some special "releases" commit. When editing the changelog, you are cleaning what should have been done at commit time before... which means your repository was left unclean until then.

Anyway, many thanks for your prompt answer, and this successfully pointed out some limitation of my strategy.

**Comment by Abdillah** [comment]
@vaab We can simplify conventions on `git commit` message. 
It will be much shorter with initial bullet point marking each operation type :
- `+` for ADDED
- `-` for REMOVED
- `#` for FIXED
- `!` for DEPRECATED
- `:` for CHANGED
- `%` for SECURITY

```
commit 20b89c9ed6c8a23e30aezy9854ad563da6cc5ca4
Author: Cool Name <email@example.com>
Date:   Thu Jul 16 21:58:15 2015 +0700

! Deprecate my_deprecated_function()
# Fix invisible column (bug #333)
: Rename my_func() to my_function() for better readability
: Move my_function() to file any.js
```

**Cool, right?**:eyeglasses: Well, but still, they don't have much semantic and need multiline commit message. 
Just an idea and sorry for bumping, at least it is open.

**Comment by simkimsia** [comment]
With regards to @vaab point that developers sometimes need to know the "why" and also pertaining to the symbols recommended by @Abdillah  may I recommend having a ? and stating the reason inside the git commit.

Also, perhaps it's a good idea to make it explicit that the ChangeLog is for external end users. and git commits are for developers working on the repo.
